File: kotlin/src/main/kotlin/com/solana/mpp/crypto/Instructions.kt
package com.solana.mpp.crypto

import com.solana.mpp.protocol.MppException

/**
 * Solana program identifiers used by the MPP charge client.
 *
 * Sourced byte-for-byte from the Rust spine at
 * `rust/src/protocol/solana.rs` (`programs` module) so the Kotlin
 * builders pick the same on-chain programs as the reference SDK.
 */
object Programs {
    const val SYSTEM_PROGRAM = "11111111111111111111111111111111"
    const val TOKEN_PROGRAM = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
    const val TOKEN_2022_PROGRAM = "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
    const val ASSOCIATED_TOKEN_PROGRAM = "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
    const val COMPUTE_BUDGET_PROGRAM = "ComputeBudget111111111111111111111111111111"
    const val MEMO_PROGRAM = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr"
}

/** Reference to a Solana account included in an instruction. */
data class AccountMeta(
    /** Base58-encoded account public key (32 byte). */
    val pubkey: String,
    /** True when the account is required to sign the transaction. */
    val isSigner: Boolean,
    /** True when the account is written to by the instruction. */
    val isWritable: Boolean,
) {
    companion object {
        /** Writable account, optionally a signer. */
        fun writable(pubkey: String, signer: Boolean = false): AccountMeta =
            AccountMeta(pubkey = pubkey, isSigner = signer, isWritable = true)

        /** Read-only account, optionally a signer. */
        fun readOnly(pubkey: String, signer: Boolean = false): AccountMeta =
            AccountMeta(pubkey = pubkey, isSigner = signer, isWritable = false)
    }
}

/** Solana program invocation. */
data class Instruction(
    val programId: String,
    val accounts: List<AccountMeta>,
    val data: ByteArray,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Instruction) return false
        return programId == other.programId &&
            accounts == other.accounts &&
            data.contentEquals(other.data)
    }

    override fun hashCode(): Int {
        var result = programId.hashCode()
        result = 31 * result + accounts.hashCode()
        result = 31 * result + data.contentHashCode()
        return result
    }
}

/**
 * Builders for the Solana on-chain programs the MPP charge intent uses.
 *
 * Every builder produces byte-for-byte identical instruction data to the
 * Rust spine. Parity oracles are documented per builder; the matching
 * Kotlin tests assert equality against the Rust source-derived golden
 * values.
 */
object Instructions {
    // ── System program ──

    /**
     * SystemProgram::transfer — moves `lamports` from `from` to `to`.
     *
     * Wire format: 4 byte little-endian discriminator `0x02 0x00 0x00 0x00`
     * followed by `lamports` as u64 little-endian.
     * Reference: `solana_system_interface::instruction::transfer`, used by
     * `rust/src/client/charge.rs:324`.
     */
    fun systemTransfer(from: String, to: String, lamports: Long): Instruction {
        require(lamports >= 0) { "lamports must be non-negative" }
        val data = ByteArray(4 + 8)
        // Discriminator 2 in little-endian u32.
        data[0] = 0x02
        encodeUInt64LE(lamports.toULong(), data, 4)
        return Instruction(
            programId = Programs.SYSTEM_PROGRAM,
            accounts = listOf(
                AccountMeta.writable(from, signer = true),
                AccountMeta.writable(to),
            ),
            data = data,
        )
    }

    // ── SPL token (and Token-2022, same wire format) ──

    /**
     * SPL token TransferChecked — moves SPL units with mint + decimals
     * cross-checked on chain.
     *
     * Wire format: 1 byte discriminator `0x0c` (12) then `amount` u64
     * little-endian then `decimals` u8.
     * Reference: `rust/src/client/charge.rs::transfer_checked_ix` at
     * line 498.
     */
    fun transferChecked(
        tokenProgram: String,
        source: String,
        mint: String,
        destination: String,
        authority: String,
        amount: Long,
        decimals: Int,
    ): Instruction {
        require(amount >= 0) { "amount must be non-negative" }
        require(decimals in 0..255) { "decimals must fit in u8" }
        val data = ByteArray(1 + 8 + 1)
        data[0] = 12
        encodeUInt64LE(amount.toULong(), data, 1)
        data[9] = decimals.toByte()
        return Instruction(
            programId = tokenProgram,
            accounts = listOf(
                AccountMeta.writable(source),
                AccountMeta.readOnly(mint),
                AccountMeta.writable(destination),
                AccountMeta.readOnly(authority, signer = true),
            ),
            data = data,
        )
    }

    // ── Associated token account ──

    /**
     * Associated Token Account program CreateIdempotent instruction.
     *
     * Wire format: 1 byte discriminator `0x01`. Accounts in fixed order:
     * payer (signer, writable), ata (writable), owner (read-only), mint
     * (read-only), system program (read-only), token program (read-only).
     * Reference: `rust/src/client/charge.rs::create_associated_token_account_idempotent`
     * at line 474.
     */
    fun createAssociatedTokenAccountIdempotent(
        payer: String,
        ata: String,
        owner: String,
        mint: String,
        tokenProgram: String,
    ): Instruction =
        Instruction(
            programId = Programs.ASSOCIATED_TOKEN_PROGRAM,
            accounts = listOf(
                AccountMeta.writable(payer, signer = true),
                AccountMeta.writable(ata),
                AccountMeta.readOnly(owner),
                AccountMeta.readOnly(mint),
                AccountMeta.readOnly(Programs.SYSTEM_PROGRAM),
                AccountMeta.readOnly(tokenProgram),
            ),
            data = byteArrayOf(0x01),
        )

    // ── Compute budget ──

    /**
     * ComputeBudget SetComputeUnitLimit.
     *
     * Wire format: 1 byte discriminator `0x02` then `units` u32 little-endian.
     * Reference: `rust/src/client/charge.rs::compute_unit_limit_ix` at
     * line 303.
     */
    fun setComputeUnitLimit(units: Int): Instruction {
        require(units >= 0) { "units must be non-negative" }
        val data = ByteArray(1 + 4)
        data[0] = 0x02
        encodeUInt32LE(units.toUInt(), data, 1)
        return Instruction(
            programId = Programs.COMPUTE_BUDGET_PROGRAM,
            accounts = emptyList(),
            data = data,
        )
    }

    /**
     * ComputeBudget SetComputeUnitPrice (price in micro-lamports).
     *
     * Wire format: 1 byte discriminator `0x03` then `microLamports` u64
     * little-endian.
     * Reference: `rust/src/client/charge.rs::compute_unit_price_ix` at
     * line 292.
     */
    fun setComputeUnitPrice(microLamports: Long): Instruction {
        require(microLamports >= 0) { "microLamports must be non-negative" }
        val data = ByteArray(1 + 8)
        data[0] = 0x03
        encodeUInt64LE(microLamports.toULong(), data, 1)
        return Instruction(
            programId = Programs.COMPUTE_BUDGET_PROGRAM,
            accounts = emptyList(),
            data = data,
        )
    }

    // ── Memo ──

    /**
     * Memo program instruction. The instruction data is the UTF-8 bytes of
     * the memo. The hard limit of 566 bytes matches the Rust spine
     * (`push_memo_instruction` in `rust/src/client/charge.rs:421`).
     */
    fun memo(memo: String): Instruction {
        val data = memo.encodeToByteArray()
        if (data.size > 566) {
            throw MppException.MemoTooLong(data.size)
        }
        return Instruction(
            programId = Programs.MEMO_PROGRAM,
            accounts = emptyList(),
            data = data,
        )
    }

    // ── Helpers ──

    internal fun encodeUInt32LE(value: UInt, out: ByteArray, offset: Int) {
        out[offset] = (value and 0xffu).toByte()
        out[offset + 1] = ((value shr 8) and 0xffu).toByte()
        out[offset + 2] = ((value shr 16) and 0xffu).toByte()
        out[offset + 3] = ((value shr 24) and 0xffu).toByte()
    }

    internal fun encodeUInt64LE(value: ULong, out: ByteArray, offset: Int) {
        var shift = 0
        for (i in 0..7) {
            out[offset + i] = ((value shr shift) and 0xffuL).toByte()
            shift += 8
        }
    }
}

File: swift/Sources/SolanaPayKit/PayCore/Instructions.swift
import Foundation

/// Solana instruction builders for the program set the MPP charge
/// intent depends on:
///
/// - SystemProgram.transfer (native SOL)
/// - SPL Token / Token-2022 transferChecked
/// - Associated Token Account create + create-idempotent
/// - ComputeBudget SetComputeUnitLimit + SetComputeUnitPrice
/// - SPL Memo (memo)
///
/// All discriminators and account orderings match the Rust spine
/// (`rust/src/client/charge.rs`) and the canonical TypeScript client
/// (`@solana-program/system`, `@solana-program/token`,
/// `@solana-program/compute-budget`). The instruction-level parity is
/// further locked by the transaction-codec parity tests in
/// `TransactionTests`: those tests already round-trip a SystemTransfer
/// + ComputeBudget + SPL transferChecked transaction byte-for-byte
/// against Rust's serialized output.
public enum Instructions {
    // MARK: System program

    /// SystemProgram::Transfer.
    /// Wire form: u32 discriminator = 2, u64 lamports (little-endian).
    public static func systemTransfer(from source: Pubkey, to destination: Pubkey, lamports: UInt64) -> SolanaInstruction {
        var data = Data()
        data.append(contentsOf: withUnsafeBytes(of: UInt32(2).littleEndian, Array.init))
        data.append(contentsOf: withUnsafeBytes(of: lamports.littleEndian, Array.init))
        return SolanaInstruction(
            programId: .systemProgram,
            accounts: [
                .writableSigner(source),
                .writable(destination),
            ],
            data: data
        )
    }

    // MARK: SPL Token / Token-2022

    /// TokenInstruction::TransferChecked.
    /// Wire form: u8 discriminator = 12, u64 amount (little-endian), u8 decimals.
    /// Account order: [source, mint (readonly), destination, authority (signer)].
    public static func splTransferChecked(
        programId: Pubkey,
        source: Pubkey,
        mint: Pubkey,
        destination: Pubkey,
        authority: Pubkey,
        amount: UInt64,
        decimals: UInt8
    ) -> SolanaInstruction {
        var data = Data([12])
        data.append(contentsOf: withUnsafeBytes(of: amount.littleEndian, Array.init))
        data.append(decimals)
        return SolanaInstruction(
            programId: programId,
            accounts: [
                .writable(source),
                .readonly(mint),
                .writable(destination),
                .readonlySigner(authority),
            ],
            data: data
        )
    }

    // MARK: Associated Token Account

    /// ATA Create (discriminator 0) or CreateIdempotent (discriminator 1).
    /// Account order matches `@solana-program/token`:
    /// [payer (writable signer), ata (writable), owner (readonly),
    ///  mint (readonly), system program (readonly), token program (readonly)].
    public static func createAssociatedTokenAccount(
        payer: Pubkey,
        ata: Pubkey,
        owner: Pubkey,
        mint: Pubkey,
        tokenProgram: Pubkey,
        idempotent: Bool
    ) -> SolanaInstruction {
        let discriminator: UInt8 = idempotent ? 1 : 0
        return SolanaInstruction(
            programId: .associatedTokenProgram,
            accounts: [
                .writableSigner(payer),
                .writable(ata),
                .readonly(owner),
                .readonly(mint),
                .readonly(.systemProgram),
                .readonly(tokenProgram),
            ],
            data: Data([discriminator])
        )
    }

    // MARK: Compute Budget

    /// SetComputeUnitLimit. Discriminator 2, u32 units (little-endian).
    public static func computeBudgetSetUnitLimit(units: UInt32) -> SolanaInstruction {
        var data = Data([2])
        data.append(contentsOf: withUnsafeBytes(of: units.littleEndian, Array.init))
        return SolanaInstruction(
            programId: .computeBudgetProgram,
            accounts: [],
            data: data
        )
    }

    /// SetComputeUnitPrice. Discriminator 3, u64 micro-lamports (little-endian).
    public static func computeBudgetSetUnitPrice(microLamports: UInt64) -> SolanaInstruction {
        var data = Data([3])
        data.append(contentsOf: withUnsafeBytes(of: microLamports.littleEndian, Array.init))
        return SolanaInstruction(
            programId: .computeBudgetProgram,
            accounts: [],
            data: data
        )
    }

    // MARK: Memo

    public static let memoMaxBytes = 566

    /// SPL Memo (memo) instruction.
    public static func memo(_ text: String) throws -> SolanaInstruction {
        let bytes = Data(text.utf8)
        guard bytes.count <= memoMaxBytes else {
            throw MppError.invalidTransaction("memo exceeds \(memoMaxBytes) bytes")
        }
        return SolanaInstruction(
            programId: .memoProgram,
            accounts: [],
            data: bytes
        )
    }
}
